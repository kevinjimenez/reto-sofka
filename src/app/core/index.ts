export { spinnerInterceptor } from './interceptors/spinner.interceptor';
export { productsResolver } from './resolvers/products.resolver';
export { ProductsService } from './services/products.service';
export { SpinnerService } from './services/spinner.service';
