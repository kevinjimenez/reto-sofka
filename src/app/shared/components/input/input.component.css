.button-container {
  display: block;
  width: 100%;
}

.button-container > .label {
  display: block;
  margin-bottom: 0.5rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 500;
}

.button-container > input {
  display: block;
  font-size: 0.875rem;
  line-height: 1.25rem;
  border-radius: 0.3rem;
  width: 100%;
  padding: 0.625rem;
  outline: none;
  box-sizing: border-box;
}

.button-container > input:focus {
  border: solid 0.15rem #7f8c8d;
}

.invalid {
  border: solid 0.15rem #e74c3c;
}


.button-container > input.invalid-focus:focus {
  border: solid 0.15rem #e74c3c;
}


.valid {
  border: solid 0.15rem #bdc3c7;
}

.error-text {
  margin-top: 0.5rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #e74c3c;
}



