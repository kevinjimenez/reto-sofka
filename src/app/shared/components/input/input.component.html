<section class="button-container">
	@if (label()) {
		<label [for]="label()" class="label">{{ label() }}</label>
	}
	<input
		[ngClass]="{
			invalid: invalidField,
			valid: !invalidField,
			'invalid-focus': invalidField
		}"
		[id]="label()"
		[formControl]="control()"
		[type]="inputType()"
		[placeholder]="placeholder()"
	/>
	@if (invalidField) {
		<p class="error-text">{{ errorMessage }}</p>
	}
</section>
