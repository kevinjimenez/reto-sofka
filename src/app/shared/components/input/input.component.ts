import { Component, input } from '@angular/core';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { CustomValiationForm } from '../../../utils/custom-validation-form';
import { NgClass } from '@angular/common';

@Component({
	selector: 'app-input',
	standalone: true,
	imports: [ReactiveFormsModule, NgClass],
	templateUrl: './input.component.html',
	styleUrl: './input.component.css'
})
export class InputComponent {
	public inputType = input<'text' | 'number' | 'date'>('text');
	public placeholder = input<string>('Search');
	public label = input<string>();
	public control = input<FormControl>(new FormControl());
	public required = input<boolean>(false);

	protected get invalidField() {
		return this.control().touched && this.control().invalid;
	}

	protected get errorMessage(): string | null {
		if (!this.control() && !this.control().errors) return null;

		return CustomValiationForm.message(this.control().errors!, this.label());
	}
}
