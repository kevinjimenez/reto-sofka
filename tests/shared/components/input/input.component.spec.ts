import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { InputComponent } from '../../../../src/app/shared/components/input/input.component';
import { ComponentRef } from '@angular/core';

describe('InputComponent', () => {
	let component: InputComponent;
	let fixture: ComponentFixture<InputComponent>;
	let componentRef: ComponentRef<InputComponent>;
	let compiled: HTMLElement;

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			imports: [InputComponent, ReactiveFormsModule]
		}).compileComponents();
	});

	beforeEach(() => {
		fixture = TestBed.createComponent(InputComponent);
		component = fixture.componentInstance;
		componentRef = fixture.componentRef;
		fixture.detectChanges();
		compiled = fixture.nativeElement;
	});

	it('should create the component', () => {
		expect(component).toBeTruthy();
	});

	it('should match snapshot', () => {
		expect(compiled).toMatchSnapshot();
	});

	it('should return not show errorMessage', () => {
		const formControlMock = new FormControl();
		componentRef.setInput('label', 'Input');
		componentRef.setInput('control', formControlMock);
		fixture.detectChanges();

		const contendErrorMessage = compiled.querySelector('.error-text');
		expect(contendErrorMessage).toBeFalsy();
		expect(compiled.querySelector('.label')).toBeTruthy();

		componentRef.setInput('label', null);
		fixture.detectChanges();

		expect(compiled.querySelector('.label')).toBeFalsy();
	});

	it('should return errorMessage', () => {
		const formControlMock = new FormControl();
		const expectedErrorMessage = 'El campo Input es requerido';
		componentRef.setInput('label', 'Input');
		componentRef.setInput('control', formControlMock);
		component.control().markAllAsTouched();
		fixture.detectChanges();

		component.control().setErrors({ required: true });
		expect(component.errorMessage).toBe(expectedErrorMessage);
		fixture.detectChanges();

		const contendErrorMessage = compiled.querySelector('.error-text');
		expect(contendErrorMessage?.textContent).toBe(expectedErrorMessage);
	});

	it('should return null when control is null', () => {
		const formControlMock = new FormControl();
		componentRef.setInput('control', formControlMock);
		fixture.detectChanges();

		componentRef.setInput('control', null);

		expect(component.errorMessage).toBeNull();
	});
});
