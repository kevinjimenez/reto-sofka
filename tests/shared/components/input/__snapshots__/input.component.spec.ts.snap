// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <section
    class="button-container"
  >
    
    <input
      class="valid ng-untouched ng-pristine ng-valid"
      id="undefined"
      placeholder="Search"
      type="text"
    />
    
  </section>
</div>
`;
